cmd/main.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/urfave/cli/v2"
+)
+
+// cmdHelp is our own help subcommand with more information
+func cmdHelp() *cli.Command {
+	c := &cli.Command{
+		Name:      "help",
+		Aliases:   []string{"h"},
+		Usage:     "Shows a list of commands or help for one command",
+		ArgsUsage: "[command]",
+		Action: func(c *cli.Context) (err error) {
+			args := c.Args()
+			if args.Present() {
+				err = cli.ShowCommandHelp(c, args.First())
+			} else {
+				err = cli.ShowAppHelp(c)
+			}
+			_, _ = fmt.Fprintf(c.App.Writer, `
+DEFAULT CONFIGURATION:
+   AppPath:    %s
+   WorkPath:   %s
+   CustomPath: %s
+   ConfigFile: %s
+
+`, setting.AppPath, setting.AppWorkPath, setting.CustomPath, setting.CustomConf)
+			return err
+		},
+	}
+	return c
+}
+
+var helpFlag = cli.HelpFlag
+
+func init() {
+	// cli.HelpFlag = nil TODO: after https://github.com/urfave/cli/issues/1794 we can use this
+}
+
+func appGlobalFlags() []cli.Flag {
+	return []cli.Flag{
+		// make the builtin flags at the top
+		helpFlag,
+		cli.VersionFlag,
+
+		// shared configuration flags, they are for global and for each sub-command at the same time
+		// eg: such command is valid: "./gitea --config /tmp/app.ini web --config /tmp/app.ini", while it's discouraged indeed
+		// keep in mind that the short flags like "-C", "-c" and "-w" are globally polluted, they can't be used for sub-commands anymore.
+		&cli.StringFlag{
+			Name:    "custom-path",
+			Aliases: []string{"C"},
+			Usage:   "Set custom path (defaults to '{WorkPath}/custom')",
+		},
+		&cli.StringFlag{
+			Name:    "config",
+			Aliases: []string{"c"},
+			Value:   setting.CustomConf,
+			Usage:   "Set custom config file (defaults to '{WorkPath}/custom/conf/app.ini')",
+		},
+		&cli.StringFlag{
+			Name:    "work-path",
+			Aliases: []string{"w"},
+			Usage:   "Set Gitea's working path (defaults to the Gitea's binary directory)",
+		},
+	}
+}
+
+func prepareSubcommandWithConfig(command *cli.Command, globalFlags []cli.Flag) {
+	command.Flags = append(append([]cli.Flag{}, globalFlags...), command.Flags...)
```

Comment:
I've never before seen an `append(append())` :eyes:

Reply:
Furthermore, isn't that equal to
```suggestion
	command.Flags = append(globalFlags, command.Flags...)
```
?

Reply:
Or will this not pass our linter?

Reply:
It needs a "copy" of the slice and it needs the "globalFlags" to be at the first.

Your suggestion would re-use the `globalFlags` slice, any change to the `command.Flags` would pollute others

Reply:
> Your suggestion would re-use the globalFlags slice

I thought an `append` in Go doesn't modify the original slice but instead returns a reference to the new slice?
If no, then why is the call `new = append(old, values...)`?
Or what do you mean?

Reply:
```
cmd1.Flags = globalFlags
cmd2.Flags = globalFlags
```

Then `cmd1.Flags[1] = newFlag` pollutes `cmd2.Flags`

Reply:
See this demo:

```go
	a := []string{"foo", "bar"}
	a1 := append(a, []string{}...)
	a2 := append(a, []string{}...)
	a1[0] = "changed"
	fmt.Printf("a1 = %v\n", a1)
	fmt.Printf("a2 = %v\n", a2)
```


```
a1 = [changed bar]
a2 = [changed bar]
```